display_list/display_list_canvas_unittests.cc

```diff
@@ -229,106 +238,209 @@ class BoundsTolerance {
+    EXPECT_TRUE(surface->readPixels(pixmap_, 0, 0));
   }
+  ~RenderResult() { free(addr_); }
+
+  int width() const { return pixmap_.width(); }
+  int height() const { return pixmap_.height(); }
+  const SkMatrix& matrix() const { return matrix_; }
+  const SkIRect& clip_bounds() const { return clip_bounds_; }
+  const uint32_t* addr32(int x, int y) const { return pixmap_.addr32(x, y); }
 
  private:
-  sk_sp<SkSurface> surface_;
   SkPixmap pixmap_;
   void* addr_ = nullptr;
+  SkMatrix matrix_;
+  SkIRect clip_bounds_;
+};
+
+#define RENDER_JOB_BASE                 \
+  int width = kTestWidth;               \
+  int height = kTestHeight;             \
+  DlColor bg = DlColor::kTransparent(); \
+  SkScalar scale = SK_Scalar1;          \
+  SkMatrix setup_matrix;                \
+  SkIRect setup_clip_bounds
+
+struct CvRenderJob {
+  RENDER_JOB_BASE;
+  CvSetup& cv_setup = kEmptyCvSetup;
+  CvRenderer& cv_render = kEmptyCvRenderer;
+  CvRenderer& cv_restore = kEmptyCvRenderer;
+  sk_sp<SkPicture> picture;
+  SkPaint setup_paint;
+
+  void render(SkCanvas* canvas) {
+    if (picture) {
+      picture->playback(canvas);
+    } else {
+      SkPaint paint;
+      cv_setup(canvas, paint);
+      setup_paint = paint;
+      setup_matrix = canvas->getTotalMatrix();
+      setup_clip_bounds = canvas->getDeviceClipBounds();
+      cv_render(canvas, paint);
+      cv_restore(canvas, paint);
+    }
+  }
+
+  sk_sp<SkPicture> get_picture() {
+    SkPictureRecorder recorder;
+    SkRTreeFactory rtree_factory;
+    SkCanvas* cv = recorder.beginRecording(kTestBounds, &rtree_factory);
+    render(cv);
+    return recorder.finishRecordingAsPicture();
+  }
+};
+
+struct DlRenderJob {
+  RENDER_JOB_BASE;
+  DlRenderer& dl_setup = kEmptyDlRenderer;
+  DlRenderer& dl_render = kEmptyDlRenderer;
+  DlRenderer& dl_restore = kEmptyDlRenderer;
+  SkScalar opacity = SK_Scalar1;
+  sk_sp<DisplayList> display_list;
+
+  void render(SkCanvas* canvas) {
+    get_display_list()->RenderTo(canvas, opacity);
+  }
+
+  sk_sp<DisplayList> get_display_list() {
+    if (!display_list) {
+      DisplayListBuilder builder(SkRect::MakeWH(width, height));
+      dl_setup(builder);
+      setup_matrix = builder.getTransform();
+      setup_clip_bounds = builder.getDestinationClipBounds().roundOut();
+      dl_render(builder);
+      dl_restore(builder);
+      display_list = builder.Build();
+    }
+    return display_list;
+  }
 };
 
 class RenderEnvironment {
  public:
-  static RenderEnvironment Make565() {
-    return RenderEnvironment(
-        SkImageInfo::Make({1, 1}, kRGB_565_SkColorType, kOpaque_SkAlphaType));
+  RenderEnvironment(const DlSurfaceProvider* provider, PixelFormat format)
+      : provider_(provider), format_(format) {
+    if (provider->supports(format)) {
+      main_surface_ =
+          provider->MakeOffscreenSurface(kTestWidth, kTestHeight, format);
+      alt_surface_ = provider->MakeOffscreenSurface(kTestWidth * 2,
```

Comment:
Can't tell how or why this alt surface exists. Can you elaborate?

Reply:
I run tests in SIZE and SIZE*2 to test dispatch-time-scaling, so the main surface is created at the normal size and the alt is created at double size.

At one point I was going to generalize this and allow the surfaces to be created dynamically depending on the request size, but it got complicated and I realized that it was far simpler to just create the 1x and 2x surfaces in the constructor.

The fields could stand to be renamed and the getSurface() method simplified after that decision.